internal/k8s/crdcontroller.go

```diff
@@ -499,7 +504,136 @@ func (c *AviController) SetupIstioCRDEventHandlers(numWorkers uint32) {
+		AddFunc: func(obj interface{}) {
+			if c.DisableSync {
+				return
+			}
+			mci := obj.(*akov1alpha1.MultiClusterIngress)
+			namespace, _, _ := cache.SplitMetaNamespaceKey(utils.ObjKey(mci))
+			key := lib.MultiClusterIngress + "/" + utils.ObjKey(mci)
+			if err := validateMultiClusterIngressObj(key, mci); err != nil {
+				utils.AviLog.Warnf("key: %s, msg: Validation of MultiClusterIngress failed: %v", key, err)
+			}
+			utils.AviLog.Debugf("key: %s, msg: ADD", key)
+			bkt := utils.Bkt(namespace, numWorkers)
+			c.workqueue[bkt].AddRateLimited(key)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			if c.DisableSync {
+				return
+			}
+			oldObj := old.(*akov1alpha1.MultiClusterIngress)
+			mci := new.(*akov1alpha1.MultiClusterIngress)
+			if !reflect.DeepEqual(oldObj.Spec, mci.Spec) {
+				namespace, _, _ := cache.SplitMetaNamespaceKey(utils.ObjKey(mci))
+				key := lib.MultiClusterIngress + "/" + utils.ObjKey(mci)
+				if err := validateMultiClusterIngressObj(key, mci); err != nil {
+					utils.AviLog.Warnf("key: %s, msg: Validation of MultiClusterIngress failed: %v", key, err)
+				}
+				utils.AviLog.Debugf("key: %s, msg: UPDATE", key)
+				bkt := utils.Bkt(namespace, numWorkers)
+				c.workqueue[bkt].AddRateLimited(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			if c.DisableSync {
+				return
+			}
+			mci, ok := obj.(*akov1alpha1.MultiClusterIngress)
+			if !ok {
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					utils.AviLog.Errorf("couldn't get object from tombstone %#v", obj)
+					return
+				}
+				mci, ok = tombstone.Obj.(*akov1alpha1.MultiClusterIngress)
+				if !ok {
+					utils.AviLog.Errorf("Tombstone contained object that is not a MultiClusterIngress: %#v", obj)
+					return
+				}
+			}
+			namespace, _, _ := cache.SplitMetaNamespaceKey(utils.ObjKey(mci))
+			key := lib.MultiClusterIngress + "/" + utils.ObjKey(mci)
+			utils.AviLog.Debugf("key: %s, msg: DELETE", key)
+			bkt := utils.Bkt(namespace, numWorkers)
+			objects.SharedResourceVerInstanceLister().Delete(key)
+			c.workqueue[bkt].AddRateLimited(key)
+		},
+	}
+	informer := lib.AKOControlConfig().CRDInformers()
+	informer.MultiClusterIngressInformer.Informer().AddEventHandler(multiClusterIngressEventHandler)
+}
+
+// SetupServiceImportEventHandlers handles setting up of ServiceImport CRD event handlers
+func (c *AviController) SetupServiceImportEventHandlers(numWorkers uint32) {
+	utils.AviLog.Infof("Setting up ServiceImport CRD Event handlers")
+
+	serviceImportEventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if c.DisableSync {
+				return
+			}
+			si := obj.(*akov1alpha1.ServiceImport)
+			namespace, _, _ := cache.SplitMetaNamespaceKey(utils.ObjKey(si))
+			key := lib.ServiceImport + "/" + utils.ObjKey(si)
+			if err := validateServiceImportObj(key, si); err != nil {
+				utils.AviLog.Warnf("key: %s, msg: Validation of ServiceImport failed: %v", key, err)
```

Comment:
Should we add the key if validation fails ?

Reply:
I have added the code to stop execution when the validation fails.